docs/writing-rules/data-flow/taint-mode.md

```diff
@@ -182,118 +500,124 @@ pattern-propagators:
+By enabling `taint_assume_safe_booleans` Semgrep will automatically sanitize Boolean expressions (when it can infer that).
 
-Remember, anything can be a sink, even the index of an array access:
+For example, comparing a tainted string against a constant string will not be considered a tainted expression:
 
-```javascript
-pattern-sinks:
-  - patterns:
-    - pattern-inside: $ARRAY[$SINK]
-    - focus-metavariable: $SINK
-```
+<iframe src="https://semgrep.dev/embed/editor?snippet=6JvzK" border="0" frameBorder="0" width="100%" height="432"></iframe>
 
-This way we tell Semgrep that we do not want arrays to be accessed with tainted indexes.
+Similarly, enabling `taint_assume_safe_numbers` Semgrep will automatically sanitize numeric expressions (when it can infer that).
 
-## Minimizing false positives
+<iframe src="https://semgrep.dev/embed/editor?snippet=oqjgX" border="0" frameBorder="0" width="100%" height="432"></iframe>
 
-Since taint mode is intra-procedural, it does not recognize what other functions do. By default, Semgrep assumes that taint can always propagate through other functions.
+You could define explicit sanitizers that clean the taint from Boolean or numeric expressions, but these options are more convenient and also more efficient. 
 
-For example, in the code below, `some_safe_function` receives tainted data as input, and to be on the safe side, Semgrep assumes that it also returns tainted data as output. As a result, a finding is produced.
+:::note
+Semgrep Pro's ability to infer types for expressions varies depending on the language. For example, in Python type annotations are not always present, and the `+` operator can also be used to concatenate strings. Semgrep also ignores the types of functions and classes coming from third-party libraries.
+
+<iframe src="https://semgrep.dev/embed/editor?snippet=zdjnn" border="0" frameBorder="0" width="100%" height="432"></iframe>
+:::
+
+### Assume tainted indexes are safe
+
+By default, Semgrep assumes that accessing an array-like object with a tainted index (that is, `obj[tainted]`) is itself tainted expression, even if the object itself is not tainted. Setting `taint_assume_safe_indexes: true` makes Semgrep assume that these expressions are safe.
+
+<iframe src="https://semgrep.dev/embed/editor?snippet=X56pj" border="0" frameBorder="0" width="100%" height="432"></iframe>
+
+### Assume function calls are safe
+
+:::note
+We refer to a function call as _opaque_ when Semgrep does not have access to its definition, in order to examine it and determine its "taint behavior" (e.g., whether the function call propagates or not any taint that comes through its inputs). In Semgrep OSS, where taint analysis is intra-procedural, all function calls are opaque. In Semgrep Pro, with [inter-procedural taint analysis](#inter-procedural-analysis-pro), an opaque function could be one coming from a third-party library.
```

Comment:
```suggestion
This documentation refers to a function call as **opaque** when Semgrep does not have access to its definition, in order to examine it and determine its "taint behavior", such as whether the function call propagates or not any taint that comes through its inputs.

In Semgrep OSS, where taint analysis is intra-procedural, all function calls are opaque.

In Semgrep Pro, with [inter-procedural taint analysis](#inter-procedural-analysis-pro), an opaque function could be one coming from a third-party library.
```

Reply:
We don't have any references at all wrt to level of "magic" that Semgrep supports per language, no? Didn't we have a reference table somewhere?

Reply:
What do you mean bye "magic"?

Reply:
Sorry, my question about magic was not directly related to opacity, but more of "what does Semgrep understand" wrt to stdlib. For example, a beginner rule-writer would not be able to infer from just reading the rule that Semgrep is able to discern that `endsWith()` returns a boolean. It's kinda subtle, but knowing that Semgrep can do that alters the way we'd write a Java rule, because we can take Semgrep's understanding of standard library functions into consideration. 

Ref: https://en.wikipedia.org/wiki/Magic_(programming)